lib/src/css_mirroring/bidi_css_gen.dart

```diff
@@ -0,0 +1,355 @@
+    if (_isRuleRemovable(removals, _originalTopLevels[iTopLevel])) {
+      _removeRuleSet(removals, iTopLevel);
+    }
+    else {
+      removals.commitDeclarations();
+      /// Add direction attribute to RuleId for direction-specific RuleSet.
+      if (mode != RetensionMode.keepBidiNeutral) {
+        _appendDirectionToRuleSet(removals, removals._topLevels[iTopLevel], targetDirection);
+      }
+    }
+  }
+
+  _editDirectives(_PendingRemovals removals, var usedDirective, List<RuleSet> originalRuleSets, List<RuleSet> flippedRuleSets, RetensionMode mode, Direction targetDirection) {
+    var _ruleStartPoints = <int>[];
+    var _ruleEndPoints = <int>[];
+    for(int iRuleSet = 0; iRuleSet < originalRuleSets.length; iRuleSet++) {
+        _storeRemovableDeclarations(removals, originalRuleSets, flippedRuleSets, mode, iRuleSet);
+        if(_isRuleRemovable(removals, originalRuleSets[iRuleSet])) {
+          var startOffset = _getRuleSetStart(originalRuleSets[iRuleSet]);
+          var endOffset = _getRuleSetEnd(usedDirective.rulesets, iRuleSet, usedDirective.span.end.offset);
+          _ruleStartPoints.add(startOffset);
+          _ruleEndPoints.add(endOffset);
+        }
+      else {
+          removals.commitDeclarations();
+          if (mode != RetensionMode.keepBidiNeutral) {
+            _appendDirectionToRuleSet(removals, usedDirective.rulesets[iRuleSet], targetDirection);
+          }
+        }
+    }
+    if(_ruleStartPoints.length == originalRuleSets.length ) { // All rules are to be deleted
+      _removeDirective(removals, usedDirective);
+    }
+    else {
+      _removeStoredRules(removals, _ruleStartPoints, _ruleEndPoints);
+    }
+  }
+
+  _storeRemovableDeclarations(_PendingRemovals removals, var originalTopLevels, var flippedTopLevels, RetensionMode mode, int iTopLevel) {
+    var originalDecls = originalTopLevels[iTopLevel].declarationGroup.declarations;
+    var flippedDecls = flippedTopLevels[iTopLevel].declarationGroup.declarations;
+
+    /// Iterate over Declarations in RuleSet and store start and end points of declarations to be removed.
+    for (int iDecl = 0; iDecl < originalDecls.length; iDecl++) {
+      if (originalDecls[iDecl] is Declaration && flippedDecls[iDecl] is Declaration) {
+        if (shouldRemoveDecl(mode, originalDecls[iDecl], flippedDecls[iDecl])) {
+          var decls = mode == RetensionMode.keepFlippedBidiSpecific ? flippedDecls : originalDecls;
+          removals.addDeclRemoval(decls[iDecl].span.start.offset, _getDeclarationEnd(removals.source, iDecl, originalDecls, decls));
+        }
+      }
+      else {
+        checkState(originalTopLevels[iTopLevel].runtimeType == flippedTopLevels[iTopLevel].runtimeType);
+      }
+    }
+  }
+
+  _appendDirectionToRuleSet(_PendingRemovals removals, RuleSet ruleSet, Direction targetDirection) {
+    removals._transaction.edit(ruleSet.span.start.offset, ruleSet.span.end.offset,
```

Comment:
instead of replacing the span with something + the span's text, how about you replace nothing by just something?

_transaction.edit(span.start.offset, span.start.offset, ':host...');